### ISSUE TITLE
Installing ngshare on UCD JupyterHub

### ISSUE THREAD

**Issue by celine168** [LibreTexts/ngshare#104]
Could I please confirm if this is the correct default `config.yaml`?
```
  2   resources:
  3     limits:
  4       cpu: 100m
  5       memory: 128Mi
  6
  7 ngshare:
  8   # You may omit this, but the token will be randomly generated.
  9   # It's recommended to specify an API token here.
 10   hub_api_token: <secret string>
 11   admins: [myusername]
 12
 13 pvc:
 14   # Amount of storage for ngshare
 15   storage: 1Gi
```
I tried installing this to a separate namespace called ngshare and it keeps spiraling into a crash loop.
```
$ kubectl logs doltish-quoll-ngshare-8d78dd89b-c97ww -n ngshare
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/ngshare/__main__.py", line 3, in <module>
    main()
TypeError: main() missing 1 required positional argument: 'argv'
```

**Comment by rkevin-arch** [comment]
That looks like an old version of ngshare (that issue should've been fixed in 1721a5b03b5f9cec369de8c91e3ca8ccf4453830). Try ngshare version `0.0.1-n313.hccd44e6`?

**Comment by celine168** [comment]
Yes, that version is installable. Thank you.

**Comment by celine168** [comment]
Hi, I get these errors after installation. Have you encountered them before?
![image](https://user-images.githubusercontent.com/22624020/83307619-ff272a80-a1b9-11ea-806a-f5f429631f0a.png)

```
$ kubectl logs  nasal-abalone-ngshare-587cc4f48-sg98c -n kube-system
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> aa00db20c10a, Init
INFO  [alembic.runtime.migration] Running upgrade aa00db20c10a -> 1921a169739b, Add file size
```

For reference, I am using this Dockerfile: https://github.com/LibreTexts/default-env/blob/ngshare/rich-default/Dockerfile

I also notice that the release name changes each time ngshare is installed. It makes upgrades a little more finicky on the user end, I was wondering if there is a benefit for it  on your end?

**Comment by rkevin-arch** [comment]
The latest installation documentation is available here: https://ngshare.readthedocs.io/en/user-guide/user_guide/install_z2jh.html Let me know if this causes more issues.

Helm2 regenerates release names unless otherwise specified, but you can just call the release `ngshare` by using `-n ngshare`. This should fix the release name changing problem.

nbgrader still needs additional configuration to see courses. This can be done by adding `c.CourseDirectory.course_id = '*'` to the `nbgrader_config.py`. After that, students can see all the courses they're in. Instructors will have to manually override this to use Formgrader one course at a time, though. More information should be available in the docs above.

**Comment by celine168** [comment]
Thanks for linking the documentation! 
I apologize, I still seem be running into an issue. It seems as though nbgrader version 0.7.0 and ngshare is installed. I am still using this as the Dockerfile/nbgrader_config.py : https://github.com/LibreTexts/default-env/tree/ngshare/rich-default 

I don't see a `Services` link show up in the Control Panel and running `nbgrader list` does give me an error.
```
$ nbgrader list
[ListApp | ERROR] Exception while loading config file /etc/jupyter/nbgrader_config.py
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 563, in _load_config_files
        config = loader.load_config()
      File "/opt/conda/lib/python3.6/site-packages/traitlets/config/loader.py", line 457, in load_config
        self._read_file_as_dict()
      File "/opt/conda/lib/python3.6/site-packages/traitlets/config/loader.py", line 489, in _read_file_as_dict
        py3compat.execfile(conf_filename, namespace)
      File "/opt/conda/lib/python3.6/site-packages/ipython_genutils/py3compat.py", line 198, in execfile
        exec(compiler(f.read(), fname, 'exec'), glob, loc)
      File "/etc/jupyter/nbgrader_config.py", line 1, in <module>
        from nbgrader.exchange import ngshare
    ImportError: cannot import name 'ngshare'
[ListApp | CRITICAL] Unwritable directory, please contact your instructor: /srv/nbgrader/exchange
[ListApp | ERROR] nbgrader list failed
```
I believe that it might be something in the way I'm writing the config file. I'm assuming that `nbgrader list` works without adding any courses yet?

**Comment by rkevin-arch** [comment]
All the stuff in `nbgrader_config.py` before line 14 is from the old installation instructions and is not needed. Only these lines are needed:

```python
from ngshare_exchange import configureExchange
c = get_config()
configureExchange(c)
c.CourseDirectory.course_id = '*'
```

`nbgrader list` should work without adding courses. It might not output anything since there are no assignments, but it should not raise any exception.

**Comment by celine168** [comment]
If we don't need to add the courses, I probably misconfigured the ngshare. I'll tackle it some more in the morning and document what I have here for now; looks like some kind of SSL error so it may not be communicating with the ngshare pod correctly...
This is what my Jhub config looks like:
```
     ngshare.py: |
        c.JupyterHub.services.append({
          'name': 'nasal-abalone-ngshare',
          'url': 'http://nasal-abalone-ngshare.kube-system.svc.local:8080',
          'api_token': <token>
```
And the error:
```
$ nbgrader list
[ListApp | ERROR] An error occurred when querying the ngshare endpoint /courses
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 485, in wrap_socket
        cnx.do_handshake()
      File "/opt/conda/lib/python3.6/site-packages/OpenSSL/SSL.py", line 1934, in do_handshake
        self._raise_ssl_error(self._ssl, result)
      File "/opt/conda/lib/python3.6/site-packages/OpenSSL/SSL.py", line 1671, in _raise_ssl_error
        _raise_current_error()
      File "/opt/conda/lib/python3.6/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
        raise exception_type(errors)
    OpenSSL.SSL.Error: [('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert internal error')]
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 672, in urlopen
        chunked=chunked,
      File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 376, in _make_request
        self._validate_conn(conn)
      File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 994, in _validate_conn
        conn.connect()
      File "/opt/conda/lib/python3.6/site-packages/urllib3/connection.py", line 360, in connect
        ssl_context=context,
      File "/opt/conda/lib/python3.6/site-packages/urllib3/util/ssl_.py", line 370, in ssl_wrap_socket
        return context.wrap_socket(sock, server_hostname=server_hostname)
      File "/opt/conda/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 491, in wrap_socket
        raise ssl.SSLError("bad handshake: %r" % e)
    ssl.SSLError: ("bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert internal error')],)",)
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
        timeout=timeout
      File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 720, in urlopen
        method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
      File "/opt/conda/lib/python3.6/site-packages/urllib3/util/retry.py", line 436, in increment
        raise MaxRetryError(_pool, url, error or ResponseError(cause))
    urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='proxy-public', port=443): Max retries exceeded with url: /services/ngshare/courses (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert internal error')],)",),))
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/ngshare_exchange/exchange.py", line 93, in ngshare_api_request
        params=params,
      File "/opt/conda/lib/python3.6/site-packages/requests/api.py", line 61, in request
        return session.request(method=method, url=url, **kwargs)
      File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 530, in request
        resp = self.send(prep, **send_kwargs)
      File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 665, in send
        history = [resp for resp in gen] if allow_redirects else []
      File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 665, in <listcomp>
        history = [resp for resp in gen] if allow_redirects else []
      File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 245, in resolve_redirects
        **adapter_kwargs
      File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 643, in send
        r = adapter.send(request, **kwargs)
      File "/opt/conda/lib/python3.6/site-packages/requests/adapters.py", line 514, in send
        raise SSLError(e, request=request)
    requests.exceptions.SSLError: HTTPSConnectionPool(host='proxy-public', port=443): Max retries exceeded with url: /services/ngshare/courses (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert internal error')],)",),))
[ListApp | CRITICAL] Failed to get courses.
[ListApp | ERROR] nbgrader list failed
```

**Comment by rkevin-arch** [comment]
ngshare_exchange shouldn't connect to proxy-public port 443 anyways. Can you run the following Python statements in the notebook server?

```python
import ngshare_exchange
e = ngshare_exchange.Exchange()
print(e.ngshare_url)
```

Also, is there a ngshare welcome page when you go to control panel -> services -> ngshare?

**Comment by celine168** [comment]
Running those statements outputs:
```
http://proxy-public/services/ngshare
```
There is no services tab in the control panel for me either.

**Comment by rkevin-arch** [comment]
Oh right, the service isn't called "ngshare" but is called "nasal-abalone-ngshare", that will change the service name that proxy-public routes. Try changing your JHub config as follows:

```
     ngshare.py: |
        c.JupyterHub.services.append({
          'name': 'ngshare',
          'url': 'http://nasal-abalone-ngshare.kube-system.svc.local:8080',
          'api_token': <token>
```

Also, the ngshare pod should not be in kube-system. I'm still not sure why yours is in there, and I can't replicate the issue even using helm2 unless I explicitly specify `--namespace kube-system`. If you do want to keep it there, make sure you can ping it from the Hub pod (maybe try `kubectl exec` into it, then try pinging that domain name just to confirm)

**Comment by celine168** [comment]
Unfortunately, changing the name back to `ngshare` leads to the same issue. I'm not so sure why the pod is in kube-system; I did put it under a different namespace earlier but forgot to specify one this time I installed ngshare.
The pod seems to be pretty strict on permissions so I can't seem to ping from inside it either.

**Comment by rkevin-arch** [comment]
Ah, I see your ngshare URL is `http://nasal-abalone-ngshare.kube-system.svc.local:8080`. It should probably be `http://nasal-abalone-ngshare.kube-system.svc.cluster.local:8080`. Maybe try that? Also try using curl inside the hub pod: `kubectl exec -it hub-whatever-it-is-called -- curl http://nasal-abalone-ngshare.kube-system.svc.local:8080`

**Comment by celine168** [comment]
```
$ kubectl exec -it hub-xxx -n staging-jhub -- curl http://nasal-abalone-ngshare.kube-system.svc.cluster.local:8080
<h1>404 Not Found</h1>
```
Quite interesting...it seemed to resolve the host just fine but didn't seem to get anything back.

```
[1] import ngshare_exchange
    e = ngshare_exchange.Exchange()
    print(e.ngshare_url)

http://proxy-public/services/ngshare
```

**Comment by celine168** [comment]
I might try to reinstall and see if that gets anywhere.

**Comment by rkevin-arch** [comment]
The 404 is normal since the service should never be accessed this way. I think I know what's going on, and I [updated the installation documentation](https://github.com/lxylxy123456/ngshare/commit/03459a5e344f86e590cdf0c998c66fd5374dccc1) to reflect some changes you need in ngshare's helm values to get it talking to JupyterHub. Alternatively, just install ngshare in the same namespace as Z2JH and these issues should go away,

**Comment by celine168** [comment]
We fixed this and ngshare is now running on the staging JupyterHub. 
These modifications were made to `nbgrader_config.py` in order for ngshare to work:
```
from ngshare_exchange import configureExchange
c = get_config()
configureExchange(c, "http://ngshare:8080/services/ngshare")
c.CourseDirectory.course_id = '*'
```

The remaining files were modified according to documentation instructions.
Our config file for the JupyterHub looks something like this:
```
hub:
  extraConfig:
    ngshare.py: |
      c.JupyterHub.services.append({
        'name': 'ngshare',
        'url': 'http://ngshare:8080',
        'api_token': '<apitoken>'})
```

And our Dockerfile looks like this:
```
RUN python3 -m pip install git+https://github.com/jupyter/nbgrader.git@5a81fd5 && \
    jupyter nbextension install --symlink --sys-prefix --py nbgrader && \
    jupyter nbextension enable --sys-prefix --py nbgrader && \
    jupyter serverextension enable --sys-prefix --py nbgrader

# Install ngshare_exchange
RUN python3 -m pip install ngshare_exchange

# Configure nbgrader
COPY nbgrader_config.py /etc/jupyter/nbgrader_config.py

RUN jupyter lab build
```

Note that for us, ngshare is deployed under the same namespace as the staging JupyterHub. 

An example of what all these files look like can be found here: https://github.com/LibreTexts/default-env/tree/ngshare/rich-default

**Comment by celine168** [comment]
@moorepants, I've added you as an admin to the staging JupyterHub. So you are able to create courses and whatnot on there using ngshare.

**Comment by rkevin-arch** [comment]
BTW, we have released ngshare and ngshare_exchange version 0.5.1. There's a bunch of bugfixes in ngshare_exchange, and the helm chart in ngshare_exchange will always output instructions with a FQDN including the namespace and also a URL for the configure_exchange function to deal with all the issues we encountered during installation. Using that version, just following the instructions should lead to a good installation without having to change stuff.
